clang/test/Sema/warn-lifetime-safety.cpp

```diff
@@ -396,6 +396,118 @@ void loan_from_previous_iteration(MyObj safe, bool condition) {
   }             // expected-note {{destroyed here}}
 }
 
+//===----------------------------------------------------------------------===//
+// Basic Definite Use-After-Return (Return-Stack-Address) (-W...permissive)
+// These are cases where the pointer is guaranteed to be dangling at the use site.
+//===----------------------------------------------------------------------===//
+
+MyObj* simple_return_stack_address(){
+  MyObj s;      
+  MyObj* p = &s; // expected-warning {{returning reference to stack allocated object}}
+  return p;      // expected-note {{returned here}}
+}
+
+MyObj* conditional_assign_unconditional_return(MyObj safe, bool c){
+  MyObj s; 
+  MyObj* p = &safe;
+  if(c){
+    p = &s;       // expected-warning {{returning reference to stack allocated object}}
+  }     
+  return p;      // expected-note {{returned here}}
+}
+
+MyObj* conditional_assign_both_branches(MyObj safe, bool c){
+
+  MyObj s;
+  MyObj* p = nullptr;
+  if (c) {
+    p = &s;     // expected-warning {{returning reference to stack allocated object}}
+  } else {
+    p = &safe;
+  }
+  return p;     // expected-note {{returned here}}
+
+}
+
+MyObj* reassign_safe_to_local(MyObj safe){
+  MyObj local;
+  MyObj* p = &safe;
+
+  p = &local;   // expected-warning {{returning reference to stack allocated object}}
+  return p;     // expected-note {{returned here}}
+}
+
+MyObj* pointer_chain_to_local(){
+  MyObj local;
+  MyObj* p1 = &local; // expected-warning {{returning reference to stack allocated object}}
+
+  MyObj* p2 = p1; 
+
+  return p2;          // expected-note {{returned here}}
+}
+
+MyObj* multiple_assign_multiple_return(MyObj safe, bool c1, bool c2){
+  MyObj local1;
+  MyObj local2;
+  MyObj* p = nullptr;
+  if(c1){
+    p = &local1;      // expected-warning {{returning reference to stack allocated object}}
+    return p;         // expected-note {{returned here}}
+  }
+  else if(c2){
+    p = &local2;      // expected-warning {{returning reference to stack allocated object}}
+    return p;         // expected-note {{returned here}}
+  }
+  p = &safe;
+  return p;
+}
+
+MyObj* multiple_assign_single_return(MyObj safe, bool c1, bool c2){
+  MyObj local1;
+  MyObj local2;
+  MyObj* p = nullptr;
+  if(c1){
+    p = &local1;     // expected-warning {{returning reference to stack allocated object}}
+  }
+  else if(c2){
+    p = &local2;     // expected-warning {{returning reference to stack allocated object}}
+  }
+  else{
+  p = &safe;
+  }
+  
+  return p;         // expected-note {{returned here}} // expected-note {{returned here}}
```

Comment:
you can specify the number of times you expect a diagnostic
`// expected-note 2 {{returned here}}`